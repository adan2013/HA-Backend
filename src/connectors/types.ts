export interface SocketMessageInterface {
  type: string
  id?: number
}

export interface EntityAttributeInterface {
  brightness?: number
  color_temp_kelvin?: number
  min_color_temp_kelvin?: number
  max_color_temp_kelvin?: number
  battery?: number
  friendly_name: string
  linkquality?: number
}

export type MessageOptions = {
  includeId: boolean
  resultCallback?: (result: any) => void
  eventCallback?: (event: any) => void
}

export type EntityState = {
  id: string
  state: string
  lastChanged: string
  lastUpdated: string
  attributes: EntityAttributeInterface
}
