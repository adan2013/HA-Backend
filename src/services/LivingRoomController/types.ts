import LightEntity from '../../entities/LightEntity'

export type BrightnessLevels = [number, number, number]

export type LightConfig = {
  levels: BrightnessLevels
  toTurnOn?: LightEntity[]
  toTurnOff?: LightEntity[]
}
