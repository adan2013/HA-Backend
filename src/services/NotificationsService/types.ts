type Priority = 'low' | 'medium' | 'high'

export type NotificationLight =
  | 'red'
  | 'yellow'
  | 'green'
  | 'blue'
  | 'purple'
  | 'redFlashing'
  | 'blueFlashing'

export type SoundType = 'scaleUp' | 'doubleBeep' | 'alarm'

export type NotificationConfig = {
  id: string
  title: string
  description: string
  priorityOrder?: Priority
  light?: NotificationLight
  sound?: SoundType
  canBeDismissed?: boolean
  ignoreDND?: boolean
}

export type NotificationPayload = {
  id: string
  title: string
  description: string
  extraInfo?: string
  priorityOrder: Priority
  light?: NotificationLight
  canBeDismissed?: boolean
  ignoreDND?: boolean
  createdAt: string
}
