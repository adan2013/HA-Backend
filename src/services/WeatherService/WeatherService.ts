import Service from '../Service'
import axios from 'axios'
import {
  AirQualityApiResponse,
  OpenWeatherOneCallApiResponse,
} from './apiTypes'
import { WeatherServiceData } from './types'
import DataCollector from '../../helpers/DataCollector'
import formatDateTime from '../../utils/formatDateTime'

class WeatherService extends Service {
  private refetchIntervalMinutes = 30
  private historicalDataHourLimit = 24
  private historicalDataSavingFrequency = 2
  private weatherEndpoint =
    'https://api.openweathermap.org/data/3.0/onecall?lat={lat}&lon={lon}&appid={apiKey}&units=metric&exclude=minutely,alerts'
  private airQualityEndpoint =
    'https://api.waqi.info/feed/{aqiStationId}?token={aqiApiKey}'
  private readonly apiKey: string
  private readonly aqiApiKey: string
  private readonly aqiStationId: string
  private readonly lat: string
  private readonly lon: string
  private readonly tempHistory: DataCollector
  private readonly windHistory: DataCollector
  private readonly pressureHistory: DataCollector
  private readonly uviHistory: DataCollector
  private readonly aqiHistory: DataCollector

  constructor() {
    super('weather')
    const limit = Math.round(
      (this.historicalDataHourLimit * (60 / this.refetchIntervalMinutes)) /
        this.historicalDataSavingFrequency,
    )
    const freq = this.historicalDataSavingFrequency
    this.tempHistory = new DataCollector('temperature', 0, limit, freq)
    this.windHistory = new DataCollector('wind', 0, limit, freq)
    this.pressureHistory = new DataCollector('pressure', 0, limit, freq)
    this.uviHistory = new DataCollector('uvi', 0, limit, freq)
    this.aqiHistory = new DataCollector('aqi', 0, limit, freq)
    this.registerHelper(this.tempHistory)
    this.apiKey = process.env['WEATHER_API_KEY'] || ''
    this.aqiApiKey = process.env['AQI_API_KEY'] || ''
    this.aqiStationId = process.env['AQI_STATION'] || ''
    this.lat = process.env['WEATHER_LAT'] || ''
    this.lon = process.env['WEATHER_LON'] || ''
    let errorMsg
    if (!this.apiKey) {
      errorMsg = 'Weather service: API key is not set'
    }
    if (!this.aqiApiKey) {
      errorMsg = 'Weather service: air quality API key is not set'
    }
    if (!this.lat || !this.lon) {
      errorMsg = 'Weather service: coordinates are not set'
    }
    if (errorMsg) {
      console.error(errorMsg)
      this.setServiceEnabled(false)
      this.setServiceStatus('Bad config', 'red')
    } else {
      setInterval(() => {
        this.fetchWeather()
      }, 60000 * this.refetchIntervalMinutes)
      this.fetchWeather()
    }
  }

  private buildApiUrl(endpointTemplate: string) {
    return endpointTemplate
      .replace('{lat}', this.lat)
      .replace('{lon}', this.lon)
      .replace('{apiKey}', this.apiKey)
      .replace('{aqiApiKey}', this.aqiApiKey)
      .replace('{aqiStationId}', this.aqiStationId)
  }

  private parseAqiValue(apiValue: number): number {
    if (apiValue) {
      if (apiValue <= 50) {
        return 1
      } else if (apiValue <= 100) {
        return 2
      } else if (apiValue <= 150) {
        return 3
      } else if (apiValue <= 200) {
        return 4
      }
      return 5
    }
    return 1
  }

  private parseWeatherData(
    weatherData: OpenWeatherOneCallApiResponse,
    airQualityData: AirQualityApiResponse,
  ): WeatherServiceData {
    const airQualityIndexValue = this.parseAqiValue(airQualityData.data.aqi)
    this.tempHistory.saveValue(weatherData.current.temp)
    this.windHistory.saveValue(weatherData.current.wind_speed * 3.6)
    this.pressureHistory.saveValue(weatherData.current.pressure)
    this.uviHistory.saveValue(weatherData.current.uvi)
    this.aqiHistory.saveValue(airQualityIndexValue)
    return {
      timestamp: new Date().getTime(),
      aqiStationName: airQualityData.data?.city?.name || '(no-name)',
      aqiStationId: airQualityData.data?.idx.toString() || '(no-id)',
      aqiStationTime: airQualityData.data?.time.iso || '(unknown)',
      current: {
        sunrise: weatherData.current.sunrise * 1000,
        sunset: weatherData.current.sunset * 1000,
        temp: weatherData.current.temp,
        feelsLike: weatherData.current.feels_like,
        pressure: weatherData.current.pressure,
        humidity: weatherData.current.humidity,
        dewPoint: weatherData.current.dew_point,
        uvi: weatherData.current.uvi,
        aqi: airQualityIndexValue,
        clouds: weatherData.current.clouds,
        visibility: weatherData.current.visibility,
        windSpeed: weatherData.current.wind_speed * 3.6,
        windDeg: weatherData.current.wind_deg,
        windGust: weatherData.current.wind_gust * 3.6,
        weather: {
          type: weatherData.current.weather[0].main,
          icon: weatherData.current.weather[0].icon,
        },
      },
      shortForecast: weatherData.hourly.map((forecast) => ({
        timestamp: forecast.dt * 1000,
        temp: forecast.temp,
        feelsLike: forecast.feels_like,
        pressure: forecast.pressure,
        uvi: forecast.uvi,
        clouds: forecast.clouds,
        visibility: forecast.visibility,
        windSpeed: forecast.wind_speed * 3.6,
        windGust: forecast.wind_gust * 3.6,
        pop: forecast.pop * 100,
        weather: {
          type: forecast.weather[0].main,
          icon: forecast.weather[0].icon,
        },
      })),
      longForecast: weatherData.daily.map((forecast) => ({
        timestamp: forecast.dt * 1000,
        sunrise: forecast.sunrise * 1000,
        sunset: forecast.sunset * 1000,
        summary: forecast.summary,
        dayTemp: forecast.temp.day,
        nightTemp: forecast.temp.night,
        minTemp: forecast.temp.min,
        maxTemp: forecast.temp.max,
        pressure: forecast.pressure,
        windSpeed: forecast.wind_speed * 3.6,
        windGust: forecast.wind_gust * 3.6,
        clouds: forecast.clouds,
        pop: forecast.pop * 100,
        uvi: forecast.uvi,
        weather: {
          type: forecast.weather[0].main,
          icon: forecast.weather[0].icon,
        },
      })),
      historicalWeather: {
        temp: this.tempHistory.getValues(),
        windSpeed: this.windHistory.getValues(),
        pressure: this.pressureHistory.getValues(),
        uvi: this.uviHistory.getValues(),
        aqi: this.aqiHistory.getValues(),
      },
    }
  }

  fetchWeather() {
    if (this.isDisabled) {
      return Promise.resolve()
    }
    const weatherUrl = this.buildApiUrl(this.weatherEndpoint)
    const airQualityUrl = this.buildApiUrl(this.airQualityEndpoint)
    return Promise.all([axios.get(weatherUrl), axios.get(airQualityUrl)])
      .then(([weatherResponse, airQualityResponse]) => {
        const weatherData =
          weatherResponse.data as OpenWeatherOneCallApiResponse
        const airQualityData = airQualityResponse.data as AirQualityApiResponse
        if (airQualityData.status !== 'ok') {
          throw new Error(`Air quality API error ${airQualityData.data}`)
        }
        const parsedData = this.parseWeatherData(weatherData, airQualityData)
        this.setServiceData(parsedData)
        this.setServiceStatus(`Weather updated at ${formatDateTime()}`, 'green')
      })
      .catch((error) => {
        this.setServiceStatus(`Error: ${error.message}`, 'red')
      })
  }
}

export default WeatherService
