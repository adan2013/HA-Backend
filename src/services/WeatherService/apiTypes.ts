type OpenWeatherState = {
  id: number
  main: string
  description: string
  icon: string
}

interface HourlyWeather {
  dt: number
  temp: number
  feels_like: number
  pressure: number
  humidity: number
  dew_point: number
  uvi: number
  clouds: number
  visibility: number
  wind_speed: number
  wind_deg: number
  wind_gust: number
  weather: OpenWeatherState[]
  pop: number
}

interface DailyWeather {
  dt: number
  sunrise: number
  sunset: number
  moonrise: number
  moonset: number
  moon_phase: number
  summary: string
  temp: {
    day: number
    min: number
    max: number
    night: number
    eve: number
    morn: number
  }
  feels_like: {
    day: number
    night: number
    eve: number
    morn: number
  }
  pressure: number
  humidity: number
  dew_point: number
  wind_speed: number
  wind_deg: number
  wind_gust: number
  weather: OpenWeatherState[]
  clouds: number
  pop: number
  uvi: number
}

export interface OpenWeatherOneCallApiResponse {
  lat: number
  lon: number
  timezone: string
  timezone_offset: number
  current: {
    dt: number
    sunrise: number
    sunset: number
    temp: number
    feels_like: number
    pressure: number
    humidity: number
    dew_point: number
    uvi: number
    clouds: number
    visibility: number
    wind_speed: number
    wind_deg: number
    wind_gust: number
    weather: OpenWeatherState[]
  }
  hourly: HourlyWeather[]
  daily: DailyWeather[]
}

export interface AirQualityApiResponse {
  status: string
  data: {
    aqi: number
    city: {
      name: string
    }
  }
}
