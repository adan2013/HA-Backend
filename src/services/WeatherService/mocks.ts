export const weatherApiMock = {
  lat: 12.34,
  lon: 56.78,
  timezone: 'Europe/Warsaw',
  timezone_offset: 3600,
  current: {
    dt: 1700848968,
    sunrise: 1700805653,
    sunset: 1700836319,
    temp: 0.78,
    feels_like: -3.73,
    pressure: 991,
    humidity: 91,
    dew_point: -0.46,
    uvi: 5,
    clouds: 75,
    visibility: 10000,
    wind_speed: 4.63,
    wind_deg: 270,
    weather: [
      {
        id: 803,
        main: 'Clouds',
        description: 'broken clouds',
        icon: '04n',
      },
    ],
  },
  hourly: [
    {
      dt: 1700848800,
      temp: 0.78,
      feels_like: -4.07,
      pressure: 991,
      humidity: 91,
      dew_point: -0.46,
      uvi: 0,
      clouds: 75,
      visibility: 10000,
      wind_speed: 5.24,
      wind_deg: 272,
      wind_gust: 11.75,
      weather: [
        {
          id: 803,
          main: 'Clouds',
          description: 'broken clouds',
          icon: '04n',
        },
      ],
      pop: 0,
    },
  ],
  daily: [
    {
      dt: 1700820000,
      sunrise: 1700805653,
      sunset: 1700836319,
      moonrise: 1700830920,
      moonset: 1700791500,
      moon_phase: 0.4,
      summary:
        'The day will start with partly cloudy with rain through the late morning hours, transitioning to partly cloudy with snow',
      temp: {
        day: 4.09,
        min: 0.01,
        max: 6.09,
        night: 0.01,
        eve: 1.15,
        morn: 4.08,
      },
      feels_like: {
        day: -1.37,
        night: -4.62,
        eve: -4.01,
        morn: -1.23,
      },
      pressure: 993,
      humidity: 74,
      dew_point: -0.33,
      wind_speed: 12.88,
      wind_deg: 255,
      wind_gust: 24.43,
      weather: [
        {
          id: 616,
          main: 'Snow',
          description: 'rain and snow',
          icon: '13d',
        },
      ],
      clouds: 81,
      pop: 1,
      rain: 2.86,
      snow: 0.95,
      uvi: 0.64,
    },
  ],
}

export const airQualityApiMock = {
  status: 'ok',
  data: {
    idx: '12345',
    aqi: 78,
    time: {
      iso: '2023-11-04T17:14:00Z',
    },
    city: {
      name: 'Warsaw, Poland',
    },
  },
}
