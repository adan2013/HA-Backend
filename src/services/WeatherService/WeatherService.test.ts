import WeatherService from './WeatherService'
import axios from 'axios'
import { airQualityApiMock, weatherApiMock } from './mocks'
import { WeatherServiceData } from './types'

const weatherApiUrl =
  'https://api.openweathermap.org/data/3.0/onecall?lat=COORDS_LAT&lon=COORDS_LON&appid=API_KEY&units=metric&exclude=minutely,alerts'
const airQualityApiUrl = 'https://api.waqi.info/feed/A12345?token=AQI_API_KEY'

jest.useFakeTimers().setSystemTime(new Date('2021-05-14T16:12:00.000'))
jest.mock('axios')
const mockedAxios = axios as jest.Mocked<typeof axios>

mockedAxios.get.mockImplementation((url) => {
  switch (url) {
    case weatherApiUrl:
      return Promise.resolve({ data: weatherApiMock })
    case airQualityApiUrl:
      return Promise.resolve({ data: airQualityApiMock })
    default:
      return Promise.reject(new Error('Wrong URL'))
  }
})

describe('WeatherService', () => {
  const OLD_ENV = process.env

  beforeEach(() => {
    jest.resetModules()
    process.env = {
      ...OLD_ENV,
      AQI_API_KEY: 'AQI_API_KEY',
      AQI_STATION: 'A12345',
      WEATHER_API_KEY: 'API_KEY',
      WEATHER_LAT: 'COORDS_LAT',
      WEATHER_LON: 'COORDS_LON',
    }
  })

  afterAll(() => {
    process.env = OLD_ENV
  })

  it('should disable sevice if one of the env variables is missing', () => {
    const initialVariables = { ...process.env }

    process.env.AQI_API_KEY = undefined
    const withoutAqiApiKey = new WeatherService()
    process.env = initialVariables

    process.env.WEATHER_API_KEY = undefined
    const withoutApiKey = new WeatherService()
    process.env = initialVariables

    process.env.WEATHER_LAT = undefined
    const withoutLat = new WeatherService()
    process.env = initialVariables

    process.env.WEATHER_LON = undefined
    const withoutLon = new WeatherService()

    expect(withoutAqiApiKey.getServiceStatus().status.enabled).toBe(false)
    expect(withoutApiKey.getServiceStatus().status.enabled).toBe(false)
    expect(withoutLat.getServiceStatus().status.enabled).toBe(false)
    expect(withoutLon.getServiceStatus().status.enabled).toBe(false)
  })

  it('should create a new instance of WeatherService with correct status', async () => {
    const weather = new WeatherService()
    await weather.fetchWeather()
    const status = weather.getServiceStatus()
    expect(status).toEqual({
      status: {
        enabled: true,
        message: 'Weather updated at 16:12 14-05-2021',
        color: 'green',
      },
      helpers: {
        'dataCollector/temperature': {
          color: 'blue',
          message: 'In queue: 1/24, Saving every 2 values',
        },
      },
    })
  })

  it('should fetch weather data', async () => {
    const weather = new WeatherService()
    await weather.fetchWeather()
    expect(mockedAxios.get).toHaveBeenCalledWith(weatherApiUrl)
    expect(mockedAxios.get).toHaveBeenCalledWith(airQualityApiUrl)
    expect(weather.getServiceData()).toMatchSnapshot()
    expect(weather.getServiceStatus().status).toEqual({
      enabled: true,
      message: 'Weather updated at 16:12 14-05-2021',
      color: 'green',
    })
  })

  it('should store historical data', async () => {
    const weather = new WeatherService()
    // it saves data every 2 values, so we need to call it 5 times
    await weather.fetchWeather()
    await weather.fetchWeather()
    await weather.fetchWeather()
    await weather.fetchWeather()
    await weather.fetchWeather()
    const historicalData = (weather.getServiceData() as WeatherServiceData)
      .historicalWeather
    expect(historicalData.temp).toHaveLength(3)
    expect(historicalData.windSpeed).toHaveLength(3)
    expect(historicalData.pressure).toHaveLength(3)
  })

  it('should set service status to error if request fails', async () => {
    mockedAxios.get.mockImplementationOnce(() => {
      return Promise.reject(new Error('ERROR_MESSAGE'))
    })
    const weather = new WeatherService()
    await weather.fetchWeather()
    expect(weather.getServiceStatus().status).toEqual({
      enabled: true,
      message: 'Error: ERROR_MESSAGE',
      color: 'red',
    })
  })

  it('should block requests if service is disabled', async () => {
    const weather = new WeatherService()
    weather.setServiceEnabled(false)
    await weather.fetchWeather()
    expect(weather.getServiceData()).toBeNull()
    expect(weather.getServiceStatus().status).toEqual({
      enabled: false,
      message: 'Ready',
      color: 'green',
    })
  })
})
