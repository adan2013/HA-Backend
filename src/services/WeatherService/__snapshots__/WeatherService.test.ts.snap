// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherService should fetch weather data 1`] = `
{
  "aqiStationId": "12345",
  "aqiStationName": "Warsaw, Poland",
  "aqiStationTime": "2023-11-04T17:14:00Z",
  "current": {
    "aqi": 2,
    "clouds": 75,
    "dewPoint": -0.46,
    "feelsLike": -3.73,
    "humidity": 91,
    "pressure": 991,
    "sunrise": 1700805653000,
    "sunset": 1700836319000,
    "temp": 0.78,
    "uvi": 5,
    "visibility": 10000,
    "weather": {
      "icon": "04n",
      "type": "Clouds",
    },
    "windDeg": 270,
    "windGust": NaN,
    "windSpeed": 16.668,
  },
  "historicalWeather": {
    "aqi": [
      2,
    ],
    "pressure": [
      991,
    ],
    "temp": [
      1,
    ],
    "uvi": [
      5,
    ],
    "windSpeed": [
      17,
    ],
  },
  "longForecast": [
    {
      "clouds": 81,
      "dayTemp": 4.09,
      "maxTemp": 6.09,
      "minTemp": 0.01,
      "nightTemp": 0.01,
      "pop": 100,
      "pressure": 993,
      "summary": "The day will start with partly cloudy with rain through the late morning hours, transitioning to partly cloudy with snow",
      "sunrise": 1700805653000,
      "sunset": 1700836319000,
      "timestamp": 1700820000000,
      "uvi": 0.64,
      "weather": {
        "icon": "13d",
        "type": "Snow",
      },
      "windGust": 87.94800000000001,
      "windSpeed": 46.368,
    },
  ],
  "shortForecast": [
    {
      "clouds": 75,
      "feelsLike": -4.07,
      "pop": 0,
      "pressure": 991,
      "temp": 0.78,
      "timestamp": 1700848800000,
      "uvi": 0,
      "visibility": 10000,
      "weather": {
        "icon": "04n",
        "type": "Clouds",
      },
      "windGust": 42.300000000000004,
      "windSpeed": 18.864,
    },
  ],
  "timestamp": 1621008720000,
}
`;
